File: src/goodbehavior/done-gate.ts
/**
 * GoodBehavior Done-Gate for Pi — pushes back on self-declared "done".
 *
 * A port of GoodBehavior's own done-gate, and it keeps that gate's principles: a heuristic nudge, not a lie detector.
 * It never judges whether the evidence is GOOD — the profile defines what real evidence is, the agent has to show it,
 * and the user confirming it is what makes work done. What the gate catches is the cheap failure: a claim with nothing
 * behind it. It is language-agnostic on purpose — it reads tool names and their order, never command contents.
 *
 * Two layers:
 *   LEXICAL    — what the message says. An honest hedge ("not done yet", "unverified") always lets it through.
 *                Verification vocabulary ("verified", "I ran", "tests pass") is only a *claim* of proof.
 *   BEHAVIORAL — what the turn actually did. Verification vocabulary is honored only if something was observed AFTER
 *                the last file change. "Edited files, ran nothing, said 'verified'" is a hollow proof.
 *
 * Three tool roles, from the profile's `done_gate` front-matter:
 *   build_tools   — using one means the turn did real work; arms the gate.
 *   mutate_tools  — tools that change files; the behavioral check asks what came after the last of these.
 *   observe_tools — running/reading/fetching the real thing.
 * `bash` belongs in build_tools AND observe_tools: it is how development work both builds and checks. That is safe
 * because the ordering is keyed on mutate_tools, not build_tools. A file written from a shell command
 * (`cat > f <<EOF`) is not seen as a change — the accepted blind spot of reading tool names, not command text.
 *
 * False-positive controls, in the order they short-circuit:
 *   (3) Activity gate  — only arm when this turn used a build tool. Pure discussion never trips it.
 *   (1) Meta escape    — talking ABOUT the gate/framework isn't a claim.
 *   (2) Assertive only — the completion word must head a short declarative clause.
 *   (0a) Hedge         — an honest downgrade always passes.
 *   (0b) Proof+observed — verification vocabulary + an observation after the last change passes. Doc-only turns (every
 *        change touched .md/.txt/…) skip the behavioral check: there is often nothing to run for prose.
 */

export interface ToolCall { name: string; input?: Record<string, unknown> }

export interface DoneGateResult {
  blocked: boolean;
  reason?: string;
  feedback?: string;
}

const DOC_EXTENSIONS = [".md", ".markdown", ".txt", ".rst", ".adoc"];

const CLAIM = /(✅|\bit'?s (now )?(done|complete|working)\b|\b(all )?(done|complete|completed|finished|shipped)\b|\bworks now\b|\bfully (working|functional)\b|\bgood to go\b|\ball set\b)/;
const HEDGE = /(not (yet|done)|isn't done|unverified|partial|in progress|pending|blocked|deferred|backlog|to verify|still to|left to do|remaining|please (check|review|confirm)|you (can )?(check|confirm))/;
const PROOF = /(verif|screenshot|i ran|ran the|test(s)? pass|passing|confirm|evidence|rendered|observed)/;
const META = /(done-gate|good ?behavior|the (gate|hook)|this hook|the regex|false (positive|trigger))/;

export class DoneGate {
  private buildTools: Set<string>;
  private observeTools: Set<string>;
  private mutateTools: Set<string>;
  private verifyHint?: string;

  // Defaults must name tools Pi actually registers: bash, edit, find, grep, ls, powershell, read, write (+ BlitzPi's
  // question, channel_post). A name outside that set never matches a real call, so it silently contributes nothing.
  // `read` is in the generic default because for non-software work reading the output/source IS the observation;
  // find/grep/ls are deliberately out — they are navigation and fire almost every turn, which would disarm the gate.
  constructor(
    buildTools: string[] = ["edit", "write", "bash"],
    observeTools: string[] = ["bash", "powershell", "read"],
    verifyHint?: string,
    mutateTools: string[] = ["edit", "write"],
  ) {
    const set = (xs: string[]) => new Set(xs.map((t) => t.toLowerCase()));
    this.buildTools = set(buildTools);
    this.mutateTools = set(mutateTools);
    // A tool that changes files cannot also be the observation of that change (fail closed on a mis-drafted profile).
    this.observeTools = new Set([...set(observeTools)].filter((t) => !this.mutateTools.has(t)));
    this.verifyHint = verifyHint?.trim() || undefined;
  }

  check(message: string, toolsCalled: ToolCall[]): DoneGateResult {
    const text = message.trim();
    if (!text) return { blocked: false };
    const t = text.toLowerCase();
    const tools = toolsCalled.map((c) => ({ name: c.name.toLowerCase(), input: c.input ?? {} }));

    // (3) Activity gate: a turn that didn't build anything can't have "finished" anything.
    if (!tools.some((c) => this.buildTools.has(c.name))) return { blocked: false };
    // (1) Meta escape: discussing the gate/framework uses words like "done" incidentally.
    if (META.test(t)) return { blocked: false };
    // (2) Assertive only: the claim must head a short clause, not lurk in a long sentence.
    if (!assertiveClaim(text)) return { blocked: false };
    // (0a) An honest hedge always passes — never punish the downgrade.
    if (HEDGE.test(t)) return { blocked: false };

    const how = this.verifyHint ?? `use ${[...this.observeTools].join(" or ")} to observe the result`;
    // (0b) Verification vocabulary passes only when the behavior backs it.
    if (PROOF.test(t)) {
      if (this.docOnlyMutations(tools) || this.observedAfterLastMutation(tools)) return { blocked: false };
      return {
        blocked: true,
        reason: "Verification claimed, but nothing was observed after the last change",
        feedback:
          "GoodBehavior done-gate — you claim verification, but this turn ran/observed NOTHING after its last file change: " +
          `the final state of the work was never exercised. Either actually observe the real thing now (${how}) and show ` +
          'the result, or downgrade the claim to "unverified / not done yet" and say what\'s left. A proof-word without ' +
          "an observation is the exact failure this method exists to stop.",
      };
    }

    return {
      blocked: true,
      reason: "Completion claimed without evidence",
      feedback:
        "GoodBehavior done-gate — you claimed completion. Before ending, self-check:\n" +
        `  1) Did you exercise the REAL thing the way its consumer would (${how}) — not just a test or your description?\n` +
        "  2) Can you SHOW the evidence (the flow firing / the output / the source-backed claim)?\n" +
        '  3) Is it user-confirmed? If not, say "not done yet" / state what\'s left — don\'t self-declare done.\n' +
        "Then either present the evidence, soften the claim, or record a learning and continue.",
    };
  }

  /** Was anything observed AFTER the last file change? A turn with no file change armed the gate through a build
   *  tool that isn't a mutation (e.g. bash) — it executed something, which is itself an observation. */
  private observedAfterLastMutation(tools: { name: string }[]): boolean {
    let last = -1;
    tools.forEach((c, i) => { if (this.mutateTools.has(c.name)) last = i; });
    if (last === -1) return true;
    return tools.slice(last + 1).some((c) => this.observeTools.has(c.name));
  }

  /** Every file change this turn touched only prose/doc files (and at least one did). */
  private docOnlyMutations(tools: { name: string; input: Record<string, unknown> }[]): boolean {
    const paths = tools
      .filter((c) => this.mutateTools.has(c.name))
      .map((c) => String(c.input.path ?? c.input.file_path ?? "").toLowerCase());
    return paths.length > 0 && paths.every((p) => DOC_EXTENSIONS.some((ext) => p.endsWith(ext)));
  }
}

/** A completion word counts only inside a short, declarative clause — not buried in prose. */
function assertiveClaim(text: string): boolean {
  return text.split(/[.!?\n]+/).some((raw) => {
    const s = raw.trim();
    return !!s && s.split(/\s+/).length <= 12 && CLAIM.test(s.toLowerCase());
  });
}

export function createDoneGate(buildTools?: string[], observeTools?: string[], verifyHint?: string, mutateTools?: string[]): DoneGate {
  return new DoneGate(buildTools, observeTools, verifyHint, mutateTools);
}

File: .claude/hooks/done-gate.py
#!/usr/bin/env python3
"""GoodBehavior done-gate — a Stop hook that pushes back on self-declared "done".

When the last assistant message makes an explicit completion claim without support, it blocks the
stop once and feeds back a short self-check. It is a heuristic nudge, not a lie detector: conservative
claim-matching to limit false positives, and a single fire per turn (the stop_hook_active guard
prevents loops).

Two layers:
  LEXICAL    — what the message says. An honest hedge ("not done yet", "unverified") always lets the
               stop through. Verification vocabulary ("verified", "I ran", "tests pass") is only a
               *claim* of proof.
  BEHAVIORAL — what the turn actually did. Verification vocabulary is honored only if something was
               actually run/observed AFTER the last file change this turn (a command, a fetch, a
               browser drive). "Edited files, ran nothing, said 'verified'" blocks: the final state
               of the work was never observed, so the proof-claim is hollow.

False-positive controls, in the order they short-circuit:
  (3) Activity gate  — only arm when THIS turn actually touched code/build (Edit/Write/Bash/...).
                       Pure discussion turns never trip the gate.
  (1) Meta escape    — talking ABOUT the gate/framework ("done-gate", "the hook") isn't a claim.
  (2) Assertive only — the completion word must head a short declarative clause, not just appear
                       somewhere in a long paragraph.
  (0a) Hedge         — an honest downgrade in the message always lets the stop through.
  (0b) Proof+observed — verification vocabulary + an observation after the last change lets it through.
       Doc-only turns (every mutation touched .md/.txt/...) are exempt from the behavioral check:
       there is often nothing to "run" for prose, so proof vocabulary suffices there.

Wire as a Stop hook in .claude/settings.json. Requires only python3.
"""
import sys, json, re

# Tools whose use means the turn did real build work (lowercased) — arms the gate.
BUILD_TOOLS = {"edit", "write", "multiedit", "notebookedit", "applypatch", "bash"}

# Tools that MUTATE files (for the behavioral check: what was the last change?).
MUTATION_TOOLS = {"edit", "write", "multiedit", "notebookedit", "applypatch"}

# Tools whose use counts as OBSERVING real behavior (running/fetching/driving something).
# Sub-agent results (task/agent) are relayed claims, not firsthand observation — excluded on purpose.
OBSERVE_TOOLS = {"bash", "webfetch"}
OBSERVE_NAME_HINTS = ("browser", "puppeteer", "playwright", "chrome")

# File extensions where "run it" usually doesn't apply — prose/docs. Doc-only turns skip the
# behavioral check (lexical proof suffices).
DOC_EXTENSIONS = (".md", ".markdown", ".txt", ".rst", ".adoc")

# Explicit completion claims (kept narrow on purpose).
CLAIM = re.compile(
    r"(✅|\bit'?s (now )?(done|complete|working)\b|\b(all )?(done|complete|completed|finished|shipped)\b"
    r"|\bworks now\b|\bfully (working|functional)\b|\bgood to go\b|\ball set\b)")

# (0a) Honest hedge / downgrade — always let the stop through; honesty must never be punished.
HEDGE = re.compile(
    r"(not (yet|done)|isn't done|unverified|partial|in progress|pending|blocked|deferred|backlog"
    r"|to verify|still to|left to do|remaining|please (check|review|confirm)|you (can )?(check|confirm))")

# (0b) Verification vocabulary — a *claim* of proof; honored only when backed by behavior.
PROOF = re.compile(
    r"(verif|screenshot|i ran|ran the|test(s)? pass|passing|confirm|evidence|rendered|observed)")

# (1) Meta-discussion about the gate/framework itself — not a real completion claim.
META = re.compile(
    r"(done-gate|good ?behavior|the (gate|hook)|this hook|stop hook|the regex"
    r"|false (positive|trigger))")

MSG_NO_EVIDENCE = (
    "GoodBehavior done-gate — you claimed completion. Before ending, self-check:\n"
    "  1) Did you exercise the REAL thing the way its consumer would (per the project's profile) — not just a test/your description?\n"
    "  2) Can you SHOW the evidence (result vs. reference; the flow firing / validation output / source-backed claim)?\n"
    "  3) Is it user-confirmed? If not, say \"not done yet\" / state what's left — don't self-declare done.\n"
    "Then either present the evidence, soften the claim, or record a learning and continue.\n")

MSG_HOLLOW_PROOF = (
    "GoodBehavior done-gate — you claim verification, but this turn ran/observed NOTHING after its last\n"
    "file change: the final state of the work was never exercised. Either actually run/observe the real\n"
    "thing now (per the project's profile) and show the result, or downgrade the claim to \"unverified /\n"
    "not done yet\" and say what's left. A proof-word without an observation is the exact failure this\n"
    "method exists to stop.\n")


def main():
    try:
        data = json.load(sys.stdin)
    except Exception:
        sys.exit(0)  # never break the session on a parse error

    # Don't re-fire on our own continuation (avoids an infinite stop loop).
    if data.get("stop_hook_active"):
        sys.exit(0)

    path = data.get("transcript_path", "")
    entries = load_entries(path)

    text = last_assistant_text(entries)
    if not text:
        sys.exit(0)
    t = text.lower()

    tools = turn_tool_uses(entries)

    # (3) Activity gate: a turn that didn't build anything can't have "finished" anything.
    if not any(name in BUILD_TOOLS for name, _ in tools):
        sys.exit(0)

    # (1) Meta escape: discussing the gate/framework uses words like "done" incidentally.
    if META.search(t):
        sys.exit(0)

    # (2) Assertive only: the claim must head a short clause, not lurk in a long sentence.
    if not assertive_claim(text):
        sys.exit(0)

    # (0a) An honest hedge always passes — never punish the downgrade.
    if HEDGE.search(t):
        sys.exit(0)

    # (0b) Verification vocabulary passes only when the behavior backs it.
    if PROOF.search(t):
        if doc_only_mutations(tools) or observed_after_last_mutation(tools):
            sys.exit(0)
        sys.stderr.write(MSG_HOLLOW_PROOF)
        sys.exit(2)

    sys.stderr.write(MSG_NO_EVIDENCE)
    sys.exit(2)  # blocks the stop; stderr is returned to the model


def assertive_claim(text):
    """A completion word counts only inside a short, declarative clause — not buried in prose."""
    for raw in re.split(r"[.!?\n]+", text):
        s = raw.strip()
        if not s:
            continue
        if len(s.split()) > 12:
            continue  # incidental mention inside a longer sentence
        if CLAIM.search(s.lower()):
            return True
    return False


def is_observe_tool(name):
    return name in OBSERVE_TOOLS or any(h in name for h in OBSERVE_NAME_HINTS)


def observed_after_last_mutation(tools):
    """Was anything run/fetched/driven AFTER the last file mutation this turn?

    A turn with no file mutations armed the gate via bash alone — it executed something, which is
    itself an observation, so it passes."""
    last_mutation = -1
    for i, (name, _) in enumerate(tools):
        if name in MUTATION_TOOLS:
            last_mutation = i
    if last_mutation == -1:
        return True
    return any(is_observe_tool(name) for name, _ in tools[last_mutation + 1:])


def doc_only_mutations(tools):
    """True if every file mutation this turn touched only prose/doc files (and at least one did)."""
    paths = []
    for name, tool_input in tools:
        if name not in MUTATION_TOOLS:
            continue
        p = ""
        if isinstance(tool_input, dict):
            p = tool_input.get("file_path") or tool_input.get("path") or tool_input.get("notebook_path") or ""
        paths.append(str(p).lower())
    if not paths:
        return False
    return all(p.endswith(DOC_EXTENSIONS) for p in paths)


def entry_role(e):
    msg = e.get("message", e)
    return e.get("role") or msg.get("role") or e.get("type")


def is_human_turn(e):
    """A real human message — not a tool_result (which the transcript also stores as role 'user')."""
    if entry_role(e) != "user":
        return False
    content = e.get("message", e).get("content", "")
    if isinstance(content, str):
        return bool(content.strip())
    if isinstance(content, list):
        return any(isinstance(b, dict) and b.get("type") == "text" and b.get("text", "").strip()
                   for b in content)
    return False


def turn_tool_uses(entries):
    """All (tool_name, tool_input) the assistant used since the last human message, in order."""
    last_human = -1
    for i, e in enumerate(entries):
        if is_human_turn(e):
            last_human = i
    out = []
    for e in entries[last_human + 1:]:
        if entry_role(e) != "assistant":
            continue
        content = e.get("message", e).get("content", "")
        if isinstance(content, list):
            for b in content:
                if isinstance(b, dict) and b.get("type") == "tool_use":
                    out.append(((b.get("name") or "").lower(), b.get("input") or {}))
    return out


def last_assistant_text(entries):
    last = ""
    for e in entries:
        if entry_role(e) != "assistant":
            continue
        content = e.get("message", e).get("content", "")
        if isinstance(content, str):
            txt = content
        elif isinstance(content, list):
            txt = " ".join(b.get("text", "") for b in content
                           if isinstance(b, dict) and b.get("type") == "text")
        else:
            txt = ""
        if txt.strip():
            last = txt
    return last


def load_entries(path):
    if not path:
        return []
    out = []
    try:
        with open(path, "r") as f:
            for line in f:
                line = line.strip()
                if not line:
                    continue
                try:
                    out.append(json.loads(line))
                except Exception:
                    continue
    except Exception:
        return []
    return out


if __name__ == "__main__":
    main()
